### PR TITLE
Handle simplified input in `query_op_helper`

### PR DESCRIPTION
**Pull request by gw000**
The inverter uses the simplifier to construct the constraints it passes out to the host VM. The VM has its own ideas about acceptable input, which don't always match up with the simplifier's output. For example, in [this case](https://github.com/osohq/oso/issues/1279) a negated comparison on a dot lookup in a partial led to the VM encountering an illegal expression when the constraint was reevaluated after one of its variables was bound.

This change adds special cases to handle rewritten dot ops in `query_op_helper`, much like we already do in `unify`. Changing the inverter not to rely on the simplifier, or changing the VM and simplifier so they agree on how to interpret expressions, would be more robust ways to fix the problem.

This PR also replaces the other `panic!`s and `assert!`s in `vm.rs` with typed errors.

PR checklist:

- [x] Added changelog entry.
- [x] Test if nested dots work (they don't, future work will address)

### CODE REVIEW

File: polar-core/src/vm.rs [hunk 1]
```diff
@@ -394,18 +399,16 @@ impl PolarVirtualMachine {
         self.stack_limit = limit;
     }
 
+    fn kb(&self) -> RwLockReadGuard<KnowledgeBase> {
```
Comment (samscott89) [inline polar-core/src/vm.rs:402]: Nice

File: polar-core/src/vm.rs [hunk 2]
```diff
@@ -523,31 +526,24 @@ impl PolarVirtualMachine {
...
-            .into());
-        }
+        use VariableState::Unbound;
         match goal {
-            Goal::LookupExternal { call_id, .. } | Goal::NextExternal { call_id, .. } => {
-                assert!(matches!(
-                    self.variable_state(self.get_call_sym(call_id)),
-                    VariableState::Unbound
-                ), "The call_id result variables for LookupExternal and NextExternal goals must be unbound.");
+            _ if self.goals.len() >= self.stack_limit =>
```
Comment (samscott89) [inline polar-core/src/vm.rs:533]: Stylistic change or functional change? If its the former, I find this more confusing.

File: polar-core/src/vm.rs [hunk 3]
```diff
@@ -571,15 +567,19 @@ impl PolarVirtualMachine {
...
-        self.choices.push(Choice {
-            alternatives,
-            bsp: self.bsp(),
-            goals: self.goals.clone(),
-            queries: self.queries.clone(),
-            trace: self.trace.clone(),
-            trace_stack: self.trace_stack.clone(),
-        });
+        if self.choices.len() >= self.stack_limit {
+            invalid_state("Too many choices.".to_string())
```
Comment (samscott89) [inline polar-core/src/vm.rs:571]: 👏

**Review by samscott89 — APPROVED** [review]
Nice! Good, pragmatic fix.

File: polar-core/src/vm.rs [hunk 4]
```diff
@@ -1758,16 +1781,68 @@ impl PolarVirtualMachine {
...
         let mut args = args.clone();
-        assert!(args.len() >= 2);
+        if args.len() < 2 {
+            return invalid_state(format!("query_op_helper: wrong arity: {}", term.to_polar()));
+        }
         let left = &args[0];
         let right = &args[1];
 
         match (left.value(), right.value()) {
+            // We may be querying a partial from the simplifier, which can contain
```
Comment (samscott89) [inline polar-core/src/vm.rs:1791]: Worth adding a TODO to rip this out once we fix the inverter?

File: polar-core/src/vm.rs (hunk 2 above)
Comment (gw000) [reply polar-core/src/vm.rs:533]: Stylistic. There's three non overlapping paths through the function so I thought it was more clear to have them all be in one block at the same indentation level

Comment (samscott89) [reply polar-core/src/vm.rs:533]: Yeah, that's fair. It seems a little weird in that the conditions are unrelated to the thing that is actually being matched on, but I don't feel that strongly about it

Comment (gw000) [reply polar-core/src/vm.rs]: I changed it to an `if/else` chain because I agree using a `match` like that is kind of odd

Comment (samscott89) [reply polar-core/src/vm.rs]: Oh I love it. Good solution

**Pull request merged**